sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TimerStateImpl.scala

```diff
@@ -0,0 +1,224 @@
+
+  private val schemaForPrefixKey: StructType = new StructType()
+    .add("key", BinaryType)
+
+  private val schemaForKeyRow: StructType = new StructType()
+    .add("key", BinaryType)
+    .add("expiryTimestampMs", LongType, nullable = false)
+
+  private val keySchemaForSecIndex: StructType = new StructType()
+    .add("expiryTimestampMs", BinaryType, nullable = false)
+    .add("key", BinaryType)
+
+  private val schemaForValueRow: StructType =
+    StructType(Array(StructField("__dummy__", NullType)))
+
+  private val keySerializer = keyExprEnc.createSerializer()
+
+  private val prefixKeyEncoder = UnsafeProjection.create(schemaForPrefixKey)
+
+  private val keyEncoder = UnsafeProjection.create(schemaForKeyRow)
+
+  private val secIndexKeyEncoder = UnsafeProjection.create(keySchemaForSecIndex)
+
+  val timerCfName = if (timeoutMode == TimeoutMode.ProcessingTime) {
+    TimerStateUtils.PROC_TIMERS_STATE_NAME
+  } else {
+    TimerStateUtils.EVENT_TIMERS_STATE_NAME
+  }
+
+  val keyToTsCFName = timerCfName + TimerStateUtils.KEY_TO_TIMESTAMP_CF
+  store.createColFamilyIfAbsent(keyToTsCFName,
+    schemaForKeyRow, numColsPrefixKey = 1,
+    schemaForValueRow, useMultipleValuesPerKey = false,
+    isInternal = true)
+
+  val tsToKeyCFName = timerCfName + TimerStateUtils.TIMESTAMP_TO_KEY_CF
+  store.createColFamilyIfAbsent(tsToKeyCFName,
+    keySchemaForSecIndex, numColsPrefixKey = 0,
+    schemaForValueRow, useMultipleValuesPerKey = false,
+    isInternal = true)
+
+  private def encodeKey(expiryTimestampMs: Long): UnsafeRow = {
+    val keyOption = ImplicitGroupingKeyTracker.getImplicitKeyOption
+    if (!keyOption.isDefined) {
+      throw StateStoreErrors.implicitKeyNotFound(keyToTsCFName)
+    }
+
+    val keyByteArr = keySerializer.apply(keyOption.get).asInstanceOf[UnsafeRow].getBytes()
+    val keyRow = keyEncoder(InternalRow(keyByteArr, expiryTimestampMs))
+    keyRow
+  }
+
+  //  We maintain a secondary index that inverts the ordering of the timestamp
+  //  and grouping key and maintains the list of (expiry) timestamps in sorted order
+  //  (using BIG_ENDIAN encoding) within RocksDB.
+  //  This is because RocksDB uses byte-wise comparison using the default comparator to
+  //  determine sorted order of keys. This is used to read expired timers at any given
+  //  processing time/event time timestamp threshold by performing a range scan.
+  private def encodeSecIndexKey(expiryTimestampMs: Long): UnsafeRow = {
+    val keyOption = ImplicitGroupingKeyTracker.getImplicitKeyOption
+    if (!keyOption.isDefined) {
+      throw StateStoreErrors.implicitKeyNotFound(tsToKeyCFName)
+    }
+
+    val keyByteArr = keySerializer.apply(keyOption.get).asInstanceOf[UnsafeRow].getBytes()
+    val bbuf = ByteBuffer.allocate(8)
+    bbuf.order(ByteOrder.BIG_ENDIAN)
+    bbuf.putLong(expiryTimestampMs)
+    val keyRow = secIndexKeyEncoder(InternalRow(bbuf.array(), keyByteArr))
```

Comment:
If I understand correctly, `bbuf.array()` is an `Array[Byte]`, which means that Spark SQL will treat this as a non-mutable type, i.e. it will encode it as a reference in the `UnsafeRow` we get back from the `UnsafeProjection`. So, then, the big-endianness of the encoding we do up above isn't really at the _start_ of the key, but rather somewhere else later on. I printed out the bytes from `keyRow` here from a test suite, and I got back:

`[0,1800000008,2000000020,cce6060000000000,0,1000000009,79656b5f74736574,31]` 

Is this behavior intentional? Would we instead want the `UnsafeRow` to look more like `[0, bbuf[0], bbuf[1], ...`.

Reply:
Yeah, good point. I see two caveats about this:

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java#L44-L58

1. Here two columns are both variable-length types - the first column will have the same relative offset and same length, so the value in reference section will all be same. The problem is in second column. The relative offset is expected to the same, but the length of second column can differ, which changes the value in reference section. It's great to have a test if we ever make the key in test as variable-length, but we should test the key with various lengths.
2. (Looks like it should not happen here but worth noting) null bits come first, which will change the orderness in overall if any of column can ever have the value of null and non-null.

That said, either we have to deal with single binary column and use its own serde format, or require state store to support orderness of prefix against some column(s), like we did for prefix scan.

Reply:
Hmm - not sure I understand. In this case - we are encoding the long type (first col) as byte array with fixed length which can never be null. So null bits,  offset(are same) and actual variable length portion (byte array) can be encoded in the provided endian-ness correctly ? Subsequent fields for the key col might support a different endian-ness and the ordering there might not be deterministic, but atleast for the case above, we did not take that as a requirement. 

Are you suggesting that we should support encoding as byte-endian for the entire unsafe row col ?

Reply:
Anish's comment makes sense.

Re: Jungtaek's comment, I'm not sure that "the length of second column can differ, which changes the value in reference section" since all timers are scoped to the same key. And since all timers always have an event-time, both the offset _and_ the size of the references in the UnsafeRow would be the same _until_ you get to the big-endian timestamp.

So, the format for any secondary index serialized key would be: [null bit set, event-time reference/size, key reference/size, event-time, key], but the first three components of that should be identical for the timers for the same key. IIUC, this secondary index approach will not work in cases where the key is variable length.

All of this is magnificently subtle, and it's not super obvious from either the comment or the code. Maybe we can improve it?

Reply:
We do not scan through timers "per key" - we scan the whole timers among keys. See getExpiredTimers().

Reply:
Oh, ok, I see that now for the get timers API. We must just be getting lucky in our test case for this, but we'd need to verify.

For the two options Jungtaek suggested, I'm not sure the latter would work, ie prefix scan against specific columns. Because it's possible that the column's actual value is not in a prefix of the encoded key. (I also might be misunderstanding your proposal.)

But the first approach could work, making your key struct type 8 literal bytes, followed by the key itself. I think that's the only way to do this correctly without relying on implementation details of the UnsafeRow?

Reply:
> require state store to support orderness of prefix against some column(s), like we did for prefix scan.

In RocksDB state store provider, we encode a bit differently for prefix scan so that it does not depend on the layout of UnsafeRow. The second option is to have a parameter to state store provider interface indicating which column(s) should be used on ordering, and encode it.

Actually it's just a matter of doing outside of state store provider vs doing inside of state store provider.

Reply:
That said, serde on ordering guarantees prefix scan for ordered columns, but prefix scan does not guarantee ordering for prefix scan columns. ordering has more strict requirement.

Reply:
Yes thats correct. Will update the PR to follow a more prefix scan like model. Thx

Reply:
I'm open to both options - OK to defer the broader change against state store API this time, till we put effort for more general use case.

Reply:
I guess we will have few applications for range scan anyway - so probably just better to add support to the underlying API ?

Reply:
Yeah otherwise let's make a quick fix (one binary column) to unblock this and introduce an API state store layer in follow-up JIRA ticket/PR. I think the change of state store API layer is worth having a separate review.

Reply:
It looks like single binary wrapped in UnsafeRow does not work - length would impact the ordering. Let's think about this a bit more.

Reply:
Done - reverted to not using the big endian ordering here. Will create a follow up for the range scan encoder